### DISCUSSION TITLE
Where does Rinja draw the line with regard to maintaining Jinja2 template compatibility

### DISCUSSION THREAD

**Discussion by cipriancraciun** [askama-rs/askama#280] [General]
I know that Rinja doesn't explicitly state anything about Jinja2 compatibility (with regard to template syntax and semantics), but in a few issues I've opened, the Jinja2 compatibility was an important factor in judging if the Rinja implementation has a bug or it just "works as expected".

**I want to highlight that I don't want to dispute the judgement of the issues in question, I just wanted to ask this broader question to have at least some guidance on the subject.**

(Personally I would actually prefer Jinja2 compatibility, given it's larger ecosystem, and the fact that one could prototype with Jinja2 in Python and then use Rinja in Rust for production.  But this is besides the point for this question.  And it's perhaps less feasible for non-trivial contexts due to the Rust / Python mismatch.)

----

Here are a few examples where Jinja2 works in one way, meanwhile Rinja differs (or vice-versa):
* for example in Jinja2, if one uses `{% extends "base.txt" %}`, then anything encountered before the statement is actually printed in the output, meanwhile in Rinja (and Askama) it is completely ignored;  (in fact Jinja2 describes where such a feature would be useful: <https://jinja.palletsprojects.com/en/stable/tricks/#null-default-fallback>));
* also, due to the above, Jinja2 allows one to strip whitespaces around `{%- extends "base.txt" -%}`, meanwhile Rinja (and Askama) fail to compile;  (I believe this should be allowed, as it's a benign usage of `{%-`, without any semantic change;)
* if one wants to output the same block twice, in Rinja (and Askama) one could use a second `{% block some_block %}{% endblock %}` and it would be rendered a second time, meanwhile Jinja2 disallows this syntax and indicates the usage of `{{ self.some_block() }}`;
* Jinja2 allows a macro caller to provide some block to the macro as in `{% call macro() %}<some-html>{% endcall %}` and the macro can then `{{ caller() }}` to include it, as opposed to Rinja (and Askama) that don't provide such a feature;  (**this would be very useful when designing HTML widgets as macros!**)

Perhaps there are more examples like this, mostly corner-cases, but I've discovered them while experimenting with other issues.

**Comment by GuillaumeGomez** [comment]
I think `rinja` should try to be as compatible with Jinja2 as possible, minus technical limitations (and even these can be revisited later on).



> * Jinja2 allows a macro caller to provide some block to the macro as in `{% call macro() %}<some-html>{% endcall %}` and the macro can then `{{ caller() }}` to include it, as opposed to Rinja (and Askama) that don't provide such a feature;  (**this would be very useful when designing HTML widgets as macros!**)

Sounds pretty neat, please open an issue so it can be added.

**Reply by cipriancraciun** [reply]
Submitted as #282 

In fact, this was something I really missed from Rinja at the moment, as it would help tremendously in developing generic-ish HTML widgets.

**Comment by seijikun** [comment]
How about the other direction:
Would it be acceptable to have more custom askama-specific syntax?

In some places, I was for example missing something like this:
```jinja
{% use crate::router::something::something::ui as ui %}
{{ ui::some_constant }}
{# instead of: #}
{{ crate::router::something::something::ui::some_constant }}
```

**Reply by GuillaumeGomez** [reply]
Problem with that is if the file containing the `use` statement is included somewhere else. Then all the new file would potentially be impacted. So even though it's convenient, the side-effects might be tricky to handle.

**Reply by seijikun** [reply]
That is true. But isn't that the case already anyway?
If I use a module or method with a relative path to the `Template`'s struct, like:

```jinja2
{{ submodule::test() }}
```

With this folder tree:
```
- src
   - my_template
      - mod.rs
      - submodule.rs
   - my_other_template
- templates
   - my_template.jinja
```
When I include `my_template.jinja` in `my_other_template`, it will not build either for all the same reasons, right?
Or did I misunderstand you?

**Reply by GuillaumeGomez** [reply]
It's exactly the same problem, except you can pass a path as an argument to go around this problem.
